lib/src/Uncertainty/Algorithm/Simulation/CrossEntropyImportanceSampling.cxx

```diff
@@ -0,0 +1,245 @@
+    {
+      const Sample blockSample(auxiliaryDistribution_.getSample(getBlockSize()));
+      auxiliaryInputSample.add(blockSample);
+      auxiliaryOutputSample.add(computeOutputSamples(blockSample));
+
+      if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+        throw InternalException(HERE) << "User stopped simulation";
+    }
+
+    // Computation of current quantile
+    currentQuantile = auxiliaryOutputSample.computeQuantile(rhoQuantile_)[0];
+    
+    // If failure probability reached, stop the adaptation
+    if (getEvent().getOperator()(currentQuantile, getEvent().getThreshold()))
+    {
+      currentQuantile = getEvent().getThreshold();
+    }
+    else
+    {
+    Indices indiceCritic(0);
+    for (UnsignedInteger i = 0; i < auxiliaryInputSample.getSize(); ++i)
+      {
+        const Bool weightBool = getEvent().getOperator()(auxiliaryOutputSample(i, 0), currentQuantile);
+        if (weightBool)
+          indiceCritic.add(i);
+      } 
+
+   // Extract the relevant sample
+   const Sample auxiliaryCriticInputSamples(auxiliaryInputSample.select(indiceCritic));
+  
+  
+   // Optimize auxiliary distribution parameters
+   Point auxiliaryDistributionParameters = optimizeAuxiliaryDistributionParameters(auxiliaryCriticInputSamples);
+  
+   // Update auxiliary Distribution Parameters
+   updateAuxiliaryDistribution(auxiliaryDistributionParameters); 
+
+    //throw InternalException(HERE) << "User stopped simulation"; 
+      
+    }
+
+    if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+      throw InternalException(HERE) << "User stopped simulation";
+
+  } // while
+
+  // Find failure sample indices
+  Indices indicesCritic(0);
+  for (UnsignedInteger i = 0; i < auxiliaryOutputSample.getSize(); ++i)
+  {
+    // Find failure Points
+    if (getEvent().getOperator()(auxiliaryOutputSample(i, 0), getEvent().getThreshold()))
+      indicesCritic.add(i);
+  } // for i
+
+  const Sample resp_sampleCritic(auxiliaryOutputSample.select(indicesCritic));
+  const Sample sampleCritic(auxiliaryInputSample.select(indicesCritic));
+
+  // Evaluate initial log PDF in parallel on failure sample
+  Sample logPDFInitCritic(initialDistribution_.computeLogPDF(sampleCritic));
+
+  // Evaluate auxiliary log PDF in parallel on failure sample
+  Sample logPDFAuxiliaryCritic(auxiliaryDistribution_.computeLogPDF(sampleCritic));
+
+  Scalar sumPdfCritic = 0.0;
+  for(UnsignedInteger i = 0; i < indicesCritic.getSize(); ++i)
+  {
+    sumPdfCritic += std::exp(logPDFInitCritic(i, 0) - logPDFAuxiliaryCritic(i, 0));
+  }
+  const Scalar failureProbability = sumPdfCritic / numberOfSample;
+  Scalar varianceCritic = 0.0;
+  for(UnsignedInteger i = 0; i < indicesCritic.getSize(); ++i)
+  {
+    const Scalar varianceCriticTemporary = std::exp(logPDFInitCritic(i, 0) - logPDFAuxiliaryCritic(i, 0)) - failureProbability;
+    varianceCritic += varianceCriticTemporary * varianceCriticTemporary;
```

Comment:
I would write:
```suggestion
    varianceCritic += pow(varianceCriticTemporary, 2);
```
because it has a richer semantic (but this does not change the value).